db.go

```diff
@@ -106,6 +111,33 @@ type Db interface {
 	// the database yet.
 	NewestSha() (sha *btcwire.ShaHash, height int64, err error)
 
+	// FetchAddrIndexTip returns the hash and block height of the most recent
+	// block which has had its address index populated. It will return
+	// ErrAddrIndexDoesNotExist along with a zero hash, and -1 if the
+	// addrindex hasn't yet been built up.
+	FetchAddrIndexTip() (sha *btcwire.ShaHash, height int64, err error)
+
+	// UpdateAddrIndexForBlock updates the stored addrindex with passed
+	// index information for a particular block height. Additionally, it
+	// will update the stored meta-data related to the curent tip of the
+	// addr index. These two operations are performed in an atomic
+	// transaction which is commited before the function returns.
+	// Addresses are indexed by the raw bytes of their base58 decoded
+	// hash160.
+	UpdateAddrIndexForBlock(blkSha *btcwire.ShaHash, height int64,
+		addrIndex *BlockAddrIndex) error
+
+	// FetchBlockBySha looks up and returns all transactions which either
+	// spend from a previously created output of the passed address, or
+	// create a new output locked to the passed address. The, `limit` parameter
+	// should be the max number of transactions to be returned. Additionally, if the
```

Comment:
The problem with unsigned size types is that when a negative number is type-switched/casted to the unsigned type, it becomes a huge positive number.  In Go this is less of a problem since the type switch is explicit but in some other languages with implicit type casting this can be the cause of many errors.  I think I'd still prefer signed here though, because it matches the types returned by `len` and `cap`, and the index for all `range` statements.

It must be documented that limit and skip must be non-negative.  As to how to handle the error cases, you can either

1. Leave them unchecked (so it's entirely the caller's responsibility)
2. Check and return an error
3. Check and panic

I would also not document which of these you use since it only matters in the case of the caller failed to satisify the invariant, and at this point they are already using the API incorrectly.